odoo/models.py

```diff
@@ -2446,6 +2453,20 @@ def _inherits_join_add(self, current_model, parent_model_name, query):
         )
         return parent_alias
 
+    def _join_calc(self, alias, fname, query):
+        field = self._fields[fname]
+        if field.type == 'many2many':
+            # LEFT JOIN field_relation ON alias.id = field_relation.field_column1
+            rel_alias = query.left_join(alias, 'id', field.relation, field.column1, 'id')
+            # LEFT JOIN field_comodel ON rel_alias.field_column2 = field_comodel.id
+            comodel = self.env[field.comodel_name]
+            comodel_table = comodel._table
+            comodel_alias = query.left_join(rel_alias, field.column2, comodel_table, 'id', field.column2)
+            # apply comodel's ir.rules
+            comodel._apply_ir_rules(query, join_alias=comodel_alias)
+            return '"%s"."%s"' % (comodel_alias, 'id')
```

Comment:
Why don't you use a subquery with only record rules, and let PG figure it out?
This is composable with the JOINs necessary for record rules.
The query would look like
```sql
SELECT "model".id
FROM "model"
LEFT JOIN "relation" AS "rel" ON "model"."id" = "rel"."left_id"
LEFT JOIN "comodel" AS "right" ON "rel"."right_id" = "right"."id" AND
    "right"."id" IN (
        -- filter rhs records with ir.rules; this may contain JOINs
        SELECT id FROM "comodel" ... WHERE ...
    )
WHERE ...
```

Reply:
I just didn't think of it, I'll give it a try. This would also my issue above with ir.rules of the comodel potentially having joins for resolving relational paths that would need to be merged/synchronized into the original query